### PR TITLE
feature: support kubernetes discovery

### PR DESCRIPTION
**Pull request by hero-zhanghao**
Ⅰ. Describe what this PR did
add kubernetes discovery

from #1076 ,Check out the previous discussion there

Ⅱ. Does this pull request fix one issue?

Ⅲ. Why don't you add test cases (unit test/integration test)?
I unit tested lockup, but I couldn't package it into a kubernetes environment for integration testing

Ⅳ. Describe how to verify it
1、Package it into a project and put it in kubernetes
2、Build a client project, use kubernetes discovery, and see if the discovery is successful

Ⅴ. Special notes for reviews
The subscribe method call is missing here and will be added to the spring cloud alibaba seata project after the release

### CODE REVIEW

**Comment by xingfudeshi** [comment]
@hero-zhanghao Hey,what i mean is you can take new branch from the develop,then do your work in the new branch,not simply submit new PR.

**Comment by hero-zhanghao** [comment]
So what I understand now is that I need to fork the project again and copy my previous code, right?

**Comment by xingfudeshi** [comment]
> So what I understand now is that I need to fork the project again and copy my previous code, right?

Yes,this will be a quick way to get everything work.

**Comment by hero-zhanghao** [comment]
@xingfudeshi I think this has been resolved, please continue to verify the code

**Comment by xingfudeshi** [comment]
> @xingfudeshi I think this has been resolved, please continue to verify the code

Good!

File: discovery/seata-discovery-kubernetes/pom.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apche.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.seata</groupId>
+        <artifactId>seata-discovery</artifactId>
+        <version>${revision}</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.seata</groupId>
```
Comment (xingfudeshi) [inline discovery/seata-discovery-kubernetes/pom.xml]: remove groupId here.

File: discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.discovery.registry.kubernetes;
+
+import io.fabric8.kubernetes.api.model.*;
```
Comment (xingfudeshi) [inline discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java]: please use explicit class type.Do not use *.

File: discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.seata.common.util.NetUtil;
+import io.seata.config.Configuration;
+import io.seata.config.ConfigurationFactory;
+import io.seata.discovery.registry.RegistryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.*;
```
Comment (xingfudeshi) [inline discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java]: please use explicit class type.Do not use *.

File: discovery/seata-discovery-kubernetes/src/test/java/io/seata/discovery/registry/kubernetes/KubernetesClientTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package io.seata.discovery.registry.kubernetes;
+
+import io.fabric8.kubernetes.api.model.*;
```
Comment (xingfudeshi) [inline discovery/seata-discovery-kubernetes/src/test/java/io/seata/discovery/registry/kubernetes/KubernetesClientTest.java]: please use explicit class type.Do not use *.

**Comment by xingfudeshi** [comment]
@hero-zhanghao Please fix  the Conflicting files.

**Comment by hero-zhanghao** [comment]
@xingfudeshi done

File: bom/pom.xml [hunk 5]
```diff
@@ -328,6 +328,17 @@
                 <artifactId>mysql-connector-java</artifactId>
                 <version>${mysql.client.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.fabric8</groupId>
+                <artifactId>kubernetes-client</artifactId>
+                <version>${fabric8.maven.plugin.version}</version>
```
Comment (slievrly) [inline bom/pom.xml]: is plugin？${fabric8.kubernetes.version}

**Comment by leizhiyuan** [comment]
please resolve the conflict

**Comment by hero-zhanghao** [comment]
@leizhiyuan @slievrly Please check if there is any problem

**Comment by hero-zhanghao** [comment]
@slievrly  Please check if the requirements are met

File: discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            }
+        }
+        return instance;
+    }
+
+
+    @Override
+    public void register(InetSocketAddress address) throws Exception {
+        if (!subscribeListener){
+            refreshCluster();
```
Comment (slievrly) [inline discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java]: How do I specify the ip and port I want to register? This information is stored in InetSocketAddress.

Comment (hero-zhanghao) [reply discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java]: In the previous PR, I discussed this with @xingfudeshi. There is no need to register in kubernetes. You can get the instance information by running in kubernetes

**Comment by hero-zhanghao** [comment]
@slievrly The conflict has been resolved. If the above answers have resolved your question, please resolve the conversation.

**Comment by hero-zhanghao** [comment]
what does dingTalk ping mean? I don't quite understand it

**Comment by xingfudeshi** [comment]
> what does dingTalk ping mean? I don't quite understand it

It's actually called DING DING in Chinese，please join the group number :23171167

**Comment by hero-zhanghao** [comment]
Ok, I'm already in there. Maybe we can communicate more quickly. It's been too long

File: discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java (hunk 6 above)
Comment (hero-zhanghao) [reply discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java]: @slievrly I want to resolve this conversation. If there are any questions, please reopen it

**Comment by hero-zhanghao** [comment]
If this is not needed, please tell me the reason and I will turn off the PR @slievrly @xingfudeshi

**Comment by slievrly** [comment]
@springrain

**Comment by oneingfw** [comment]
Are there anyone focus on this feature?

**Comment by funky-eyes** [comment]
Due to code conflicts and age, if you are still interested in participating in the community, please resubmit this pr to the 2.x branch after resolving the conflict.
由于代码冲突以及年代久远，如果你还有兴趣参与到社区，请解决完冲突后重新请提交这份pr至2.x分支中

**Pull request closed without merging**